### ISSUE TITLE
Grouping entities

### ISSUE THREAD

**Issue by Yelnats321** [Yelnats321/EntityPlus#9]
As it stands for_each will iterate over the smallest component/tag provided, however this means that if the intersection of the provided components/tags is very small compared to the smallest component/tags, we are doing a lot of unnecessary work. It may be helpful for the user to provide groups of entities they plan on iterating over in the future to reduce the cost of this iteration.

As an example, say there are 50 entities, 20 of which have only tag A, 25 which have only tag B, and 5 which have both A and B. If we did `for_each<A, B>()` we would iterate over 20 entities (the ones with tag A) and then compare the bit signatures of the entities to match AB. Instead the user may tell us that they will use a grouping of AB regularly and so now we have a separate list of all entities with AB that we update with changes, and now we only have to iterate over the 5 AB entities.

**Comment by Yelnats321** [comment]
The branch [entity_groups](/Yelnats321/EntityPlus/tree/entity_groups) for this change. It adds grouping as well as group handles and simplifies the default group creation because it generalizes this concept.

**Comment by Yelnats321** [comment]
With the  current implementation, `get_smallest_idx()` (which should be renamed to `get_smallest_container()`) needs to sort every time, costing `O(k log k)`, where `k` is the number of groupings. Also currently `add/remove_bit()` costs `O(k log n)` as well, because of the `find()`. This was kind of not the case before, because `k` used to be a constant. What can be done instead is to keep a `flat_set` of the groupings instead, sorted by size. Then `get_smallest_idx()` will be `O(log k)`. `entity_grouping::destroy()` will suffer but that doesn't really matter. However, we will need to adjust the `add/remove_bit()` code as erasing/adding to a grouping may alter the ordering, so we need to be careful. This does mean that `add/remove_bit()` will now cost `O(k(k + log n))` because we will need to rotate if our location changes. This is very likely to be ok, since adding/removing should be done more rarely than iterating, and because the amount of groupings should be small.

Actually, we don't even need to rotate, just swap two groupings because we just need weak ordering, so the cost would really be `O(k(log k + log n)` which will definitely be fast enough.

The other works that needs to be done is more tests and added documentation.

Also, `for_each` may benefit from knowing if a grouping wholly represents the entities to be iterated, because then the branch prediction can be better, although this is speculation.

**Comment by Yelnats321** [comment]
`for_each` now takes into account if the grouping is the wholly searched for group.

There was a bug with the finding the smallest container which could find a container that was too restrictive. For example if we had a grouping for entities ABCD and we searched for entities with A and B, the smallest grouping could be the ABCD grouping because it is smaller in size. However this grouping does not encompass all entities with AB. This has now been fixed.

It does not seem to be possible to keep the groupings sorted, because the requirement for a binary search would be that the groupings are partitioned with regards to the searched for grouping, which is impossible in the general case, so this idea was abandoned.
